<!DOCTYPE html>
<html class="writer-html5" lang="en">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Beginner’s tutorial &mdash; chloroscan  documentation</title>
      <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="../_static/copybutton.css?v=76b2166b" />
      <link rel="stylesheet" type="text/css" href="../_static/graphviz.css?v=eafc0fe6" />

  
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="../_static/jquery.js?v=5d32c60e"></script>
        <script src="../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js?v=b3ba4146"></script>
        <script src="../_static/doctools.js?v=888ff710"></script>
        <script src="../_static/sphinx_highlight.js?v=4825356b"></script>
        <script src="../_static/clipboard.min.js?v=a7894cd8"></script>
        <script src="../_static/copybutton.js?v=f281be69"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="Command Line Interface" href="cli.html" />
    <link rel="prev" title="ChloroScan: A metagenomic workflow to recover chloroplast genomes." href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            chloroscan
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">ChloroScan: A metagenomic workflow to recover chloroplast genomes.</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Beginner’s tutorial</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#data-input-and-configuration">Data Input and configuration</a></li>
<li class="toctree-l2"><a class="reference internal" href="#explaining-the-non-required-arguments">Explaining the non-required arguments</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#corgi-settings">1. CORGI settings</a></li>
<li class="toctree-l3"><a class="reference internal" href="#binning-module-settings">2. Binning module settings</a></li>
<li class="toctree-l3"><a class="reference internal" href="#homology-based-taxonomy-prediction-by-cat">3. Homology-based taxonomy prediction by CAT</a></li>
<li class="toctree-l3"><a class="reference internal" href="#other-settings">4. Other settings</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="cli.html">Command Line Interface</a></li>
<li class="toctree-l1"><a class="reference internal" href="credits.html">Credits</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">chloroscan</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Beginner’s tutorial</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/main/beginner_tutorial.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="beginner-s-tutorial">
<h1>Beginner’s tutorial<a class="headerlink" href="#beginner-s-tutorial" title="Permalink to this heading"></a></h1>
<section id="data-input-and-configuration">
<h2>Data Input and configuration<a class="headerlink" href="#data-input-and-configuration" title="Permalink to this heading"></a></h2>
<p>The pip version of ChloroScan accepts following arguments:</p>
<img alt="../_images/chloroscan_cli.png" src="../_images/chloroscan_cli.png" />
<p>Among these arguments, only assembly path, depth text, alignment folder and a batch name are required for ChloroScan to run.</p>
<p>Please prepare your assembly in unziped fasta format to allow for processing.</p>
<p>The sequence alignment folder should contain sorted.bam files to map each contig’s sequence coverage in assembly.</p>
<p>An example command for running ChloroScan if depth text file is available:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chloroscan<span class="w"> </span>run<span class="w"> </span>--Inputs-assembly-path<span class="w"> </span>assembly.fasta<span class="w"> </span>--Inputs-depth-txt<span class="w"> </span>depth.txt<span class="w"> </span>--Inputs-batch-name<span class="w"> </span>BATCH_NAME<span class="w"> </span>--outputdir<span class="w"> </span>OUTPUT_DIR<span class="w"> </span>--cores<span class="w"> </span>N_CORES
</pre></div>
</div>
<p>An example command for running ChloroScan if depth text file is not available:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>chloroscan<span class="w"> </span>run<span class="w"> </span>--Inputs-assembly-path<span class="w"> </span>assembly.fasta<span class="w"> </span>--Inputs-batch-name<span class="w"> </span>BATCH_NAME<span class="w"> </span>--outputdir<span class="w"> </span>OUTPUT_DIR<span class="w"> </span>--alignment-folder<span class="w"> </span>ALIGNMENT_FOLDER
</pre></div>
</div>
</section>
<section id="explaining-the-non-required-arguments">
<h2>Explaining the non-required arguments<a class="headerlink" href="#explaining-the-non-required-arguments" title="Permalink to this heading"></a></h2>
<p>Some parameters are not required, but changing them can impact the outcomes and running processes, here are some instructions.</p>
<section id="corgi-settings">
<h3>1. CORGI settings<a class="headerlink" href="#corgi-settings" title="Permalink to this heading"></a></h3>
<p>CORGI is responsible for contig filtering to extract chloroplast contigs. The default device of CORGI is GPU.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--corgi-settings-min-length</span></code>: Minimum length of contigs to be considered for chloroplast contigs. Default is 500.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--corgi-settings-pthreshold</span></code>: P-value threshold for filtering contigs. Default is 0.90. Bigger value means more stringent filtering.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--corgi-settings-save-filtering</span></code>: Whether to save the filtered contigs using CORGI. Default is False and the customized python scripts in chloroscan could achieve quite fast writing speed.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--corgi-settings-batch-size</span></code>: number of contigs to be processed in one batch. Default is 1. Smaller batch size can reduce memory usage and elevate speed.</p></li>
</ul>
</section>
<section id="binning-module-settings">
<h3>2. Binning module settings<a class="headerlink" href="#binning-module-settings" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-universal-length-cutoff</span></code>: The length cutoff for all contigs that enters selection regardless of presenting marker genes. Default is 1500.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-clustering-epsilon_range</span></code>: The range of epsilon for HDBSCAN clustering. Default is “0.250,0.000”. Larger range takes more time to finish the clustering.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-clustering-hdbscan-min_samples</span></code>: The minimum number of samples in a cluster. Default is 1,4,7,10. Picking smaller values can lead to more clusters but cluster contents may be spurious.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-bin-quality-min_completeness</span></code>: Minimum completeness for MAGs to enter selection. Default is 72.5.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-bin-quality-starting-completeness</span></code>: Starting completeness for MAGs to enter selection. Default is 92.5, note that the range between starting and min completeness affects the quality selections of bins.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--binning-bin-quality-purity</span></code>: Purity cutoff for MAGs to enter selection. Default is 95.</p></li>
</ul>
</section>
<section id="homology-based-taxonomy-prediction-by-cat">
<h3>3. Homology-based taxonomy prediction by CAT<a class="headerlink" href="#homology-based-taxonomy-prediction-by-cat" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">CAT-database</span></code>: The database used for taxonomy prediction by CAT/BAT, ensure this is an absolute path.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">CAT-taxonomy</span></code>: The taxonomy dump file for CAT to use. Ensure this path is an absolute path.</p></li>
</ul>
</section>
<section id="other-settings">
<h3>4. Other settings<a class="headerlink" href="#other-settings" title="Permalink to this heading"></a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">outputdir</span></code>: The output directory for the workflow’s results, usually named after batch name.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">tmpdir</span></code>: Temporary directory for storing intermediate files. Default is /tmp.</p></li>
</ul>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Currently the pip version of chloroscan recommends to use a snakemake virtual environment with snakemake=6.15.5 installed for binny’s module to use.
Otherwise the binny module might fail while running.</p>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="ChloroScan: A metagenomic workflow to recover chloroplast genomes." accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="cli.html" class="btn btn-neutral float-right" title="Command Line Interface" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2024–2025, Yuhao Tong.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>